File: sdks/dotnet/ChioMiddleware/src/ChioSidecarClient.cs
// Chio sidecar HTTP client for .NET.
//
// Communicates with the Chio Rust kernel running as a localhost sidecar.
// Sends evaluation requests over HTTP and returns signed receipts.

using System.Net.Http.Json;
using System.Text.Json;

namespace Backbay.Arc;

/// <summary>
/// Exception thrown when the Chio sidecar is unreachable or returns an error.
/// </summary>
public class ChioSidecarException : Exception
{
    public string Code { get; }
    public int? StatusCode { get; }

    public ChioSidecarException(string code, string message, int? statusCode = null)
        : base(message)
    {
        Code = code;
        StatusCode = statusCode;
    }
}

/// <summary>
/// Chio sidecar client. Sends evaluation requests to the Rust kernel.
/// </summary>
public class ChioSidecarClient : IDisposable
{
    public const string DefaultSidecarUrl = "http://127.0.0.1:9090";

    private readonly string _baseUrl;
    private readonly HttpClient _httpClient;
    private readonly JsonSerializerOptions _jsonOptions;

    public ChioSidecarClient(string? baseUrl = null, int timeoutSeconds = 5)
    {
        _baseUrl = (baseUrl ?? Environment.GetEnvironmentVariable("CHIO_SIDECAR_URL") ?? DefaultSidecarUrl)
            .TrimEnd('/');
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(timeoutSeconds)
        };
        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
        };
    }

    /// <summary>
    /// Evaluate an HTTP request against the Chio kernel.
    /// </summary>
    public async Task<EvaluateResponse> EvaluateAsync(ChioHttpRequest request, string? capabilityToken = null)
    {
        HttpResponseMessage response;
        try
        {
            using var message = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/chio/evaluate")
            {
                Content = JsonContent.Create(request, options: _jsonOptions)
            };
            if (!string.IsNullOrWhiteSpace(capabilityToken))
            {
                message.Headers.Add("X-Chio-Capability", capabilityToken);
            }
            response = await _httpClient.SendAsync(message);
        }
        catch (Exception ex)
        {
            throw new ChioSidecarException(
                ChioErrorCodes.SidecarUnreachable,
                $"Failed to reach Chio sidecar at {_baseUrl}: {ex.Message}"
            );
        }

        if (!response.IsSuccessStatusCode)
        {
            var body = await response.Content.ReadAsStringAsync();
            throw new ChioSidecarException(
                ChioErrorCodes.EvaluationFailed,
                $"Sidecar returned {(int)response.StatusCode}: {body}",
                (int)response.StatusCode
            );
        }

        var result = await response.Content.ReadFromJsonAsync<EvaluateResponse>(_jsonOptions);
        return result ?? throw new ChioSidecarException(
            ChioErrorCodes.EvaluationFailed,
            "Sidecar returned null response"
        );
    }

    /// <summary>
    /// Verify a receipt signature against the sidecar.
    /// </summary>
    public async Task<bool> VerifyReceiptAsync(HttpReceipt receipt)
    {
        try
        {
            var response = await _httpClient.PostAsJsonAsync(
                $"{_baseUrl}/chio/verify",
                receipt,
                _jsonOptions
            );

            if (!response.IsSuccessStatusCode)
                return false;

            var result = await response.Content.ReadFromJsonAsync<Dictionary<string, object>>(_jsonOptions);
            return result?.TryGetValue("valid", out var valid) == true && valid is JsonElement elem && elem.GetBoolean();
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    /// Health check for the sidecar.
    /// </summary>
    public async Task<bool> HealthCheckAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_baseUrl}/chio/health");
            return response.IsSuccessStatusCode;
        }
        catch
        {
            return false;
        }
    }

    public void Dispose()
    {
        _httpClient.Dispose();
        GC.SuppressFinalize(this);
    }
}

File: sdks/jvm/chio-spring-boot/src/main/kotlin/io/backbay/chio/ChioSidecarClient.kt
/**
 * Chio sidecar HTTP client for JVM.
 *
 * Communicates with the Chio Rust kernel running as a localhost sidecar.
 * Sends evaluation requests over HTTP and returns signed receipts.
 */
package io.backbay.chio

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/** Exception thrown when the Chio sidecar is unreachable or returns an error. */
class ChioSidecarException(
    val code: String,
    override val message: String,
    val statusCode: Int? = null,
) : RuntimeException(message)

/** Chio sidecar client. Sends evaluation requests to the Rust kernel. */
class ChioSidecarClient(
    private val baseUrl: String = DEFAULT_SIDECAR_URL,
    private val timeoutSeconds: Long = 5,
) {
    companion object {
        const val DEFAULT_SIDECAR_URL = "http://127.0.0.1:9090"
    }

    private val httpClient: HttpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(timeoutSeconds))
        .build()

    private val objectMapper: ObjectMapper = jacksonObjectMapper()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    /** Evaluate an HTTP request against the Chio kernel. */
    fun evaluate(request: ChioHttpRequest, capabilityToken: String? = null): EvaluateResponse {
        val body = objectMapper.writeValueAsString(request)
        val requestBuilder = HttpRequest.newBuilder()
            .uri(URI.create("$baseUrl/chio/evaluate"))
            .header("Content-Type", "application/json")
            .timeout(Duration.ofSeconds(timeoutSeconds))
            .POST(HttpRequest.BodyPublishers.ofString(body))
        if (!capabilityToken.isNullOrBlank()) {
            requestBuilder.header("X-Chio-Capability", capabilityToken)
        }
        val httpRequest = requestBuilder.build()

        val response = try {
            httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())
        } catch (e: Exception) {
            throw ChioSidecarException(
                code = ChioErrorCodes.SIDECAR_UNREACHABLE,
                message = "failed to reach Chio sidecar at $baseUrl: ${e.message}",
            )
        }

        if (response.statusCode() >= 400) {
            throw ChioSidecarException(
                code = ChioErrorCodes.EVALUATION_FAILED,
                message = "sidecar returned ${response.statusCode()}: ${response.body()}",
                statusCode = response.statusCode(),
            )
        }

        return objectMapper.readValue(response.body())
    }

    /** Verify a receipt signature against the sidecar. */
    fun verifyReceipt(receipt: HttpReceipt): Boolean {
        val body = objectMapper.writeValueAsString(receipt)
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseUrl/chio/verify"))
            .header("Content-Type", "application/json")
            .timeout(Duration.ofSeconds(timeoutSeconds))
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build()

        return try {
            val response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())
            if (response.statusCode() != 200) return false
            val result: Map<String, Any> = objectMapper.readValue(response.body())
            result["valid"] == true
        } catch (_: Exception) {
            false
        }
    }

    /** Health check for the sidecar. */
    fun healthCheck(): Boolean {
        val httpRequest = HttpRequest.newBuilder()
            .uri(URI.create("$baseUrl/chio/health"))
            .timeout(Duration.ofSeconds(timeoutSeconds))
            .GET()
            .build()

        return try {
            val response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())
            response.statusCode() == 200
        } catch (_: Exception) {
            false
        }
    }
}
